### PR TITLE
Add EmailLoader

### LINKED ISSUE #126: Add `EmailLoader`
Move email loading logic from the `EmailClient` tool in `griptape-tools` to the `EmailLoader`.

### PR DESCRIPTION
**Pull request by dylanholmes**
### Notes

Resolves #126:
- Adds EmailLoader with unit tests.
- Improves unit test coverage of the EmailClient tool.

### Testing

Unit tests with 100% branch and line coverage for the email loader and client.
```
poetry run pytest \
  tests/unit/tools/test_email_client.py --cov=griptape/tools/email_client/ \
  tests/unit/loaders/test_email_loader.py --cov=griptape/loaders
```

Relevant coverage:
```
Name                                      Stmts   Miss Branch BrPart  Cover
---------------------------------------------------------------------------
...
griptape/loaders/email_loader.py             53      0     14      0   100%
...
griptape/tools/email_client/__init__.py       0      0      0      0   100%
griptape/tools/email_client/tool.py          53      0     14      0   100%
---------------------------------------------------------------------------
TOTAL                                       297     74    106      4    67%
```

Manually tested email client tool with personal gmail account for send and retrieve.

> Note: I considered adding an integration test with a locally running mail server. Adding one for the send functionality was pretty easy (without using SSL that is) since there are many smtp python implementations. I couldn't find a simple/lightweight imap server implementation for testing the retrieve functionality.
>
> I think a better approach for integration testing would be to use a griptape maintained email address specifically for testing the email client, but this is out of scope for this PR imo.

### CODE REVIEW

File: griptape/loaders/email_loader.py [hunk 1]
```diff
@@ -0,0 +1,81 @@
...
+    """Label to retrieve emails from such as 'INBOX' or 'SENT'"""
+
+    key: Optional[str] = field(default=None, kw_only=True)
+    """Optional key for filtering such as 'FROM' or 'SUBJECT'"""
+
+    search_criteria: Optional[str] = field(default=None, kw_only=True)
+    """Optional search criteria to filter emails by key"""
+
+    max_count: Optional[int] = field(default=None, kw_only=True)
+    """Optional max email count"""
```
Comment (collindutter) [inline griptape/loaders/email_loader.py:27]: Can you please update the class doc string to match [Google's docstring style](https://google.github.io/styleguide/pyguide.html#384-classes)? It's what our auto generated reference docs expects.

File: griptape/loaders/email_loader.py [hunk 2]
```diff
@@ -0,0 +1,81 @@
...
+                        )
+
+                client.close()
+
+                return list_artifact
+        except Exception as e:
+            logging.error(e)
+            return ErrorArtifact(f'error retrieving email: {e}')
+
+def count_messages(message_numbers: bytes):
```
Comment (collindutter) [inline griptape/loaders/email_loader.py:80]: Change to protected method.

File: griptape/tools/email_client/tool.py [hunk 3]
```diff
@@ -1,56 +1,70 @@
...
-        smtp_host: The hostname or url of the SMTP server (smtp.gmail.com).
-        smtp_port: The port name of the SMTP server (465).
-        smtp_use_ssl: Should EmailClient use SSL when sending.
-        smtp_user: Setting this will override whatever is set as the username parameter for SMTP activities.
-        smtp_password: Setting this will override whatever is set as the username parameter for SMTP activities.
-        imap_url: The hostname or url of the SMTP server (imap.gmail.com).
-        imap_user: Setting this will override whatever is set as the username parameter for IMAP activities.
-        imap_password: Setting this will override whatever is set as the username parameter for IMAP activities.
-    """
+    """Tool for working with email"""
```
Comment (collindutter) [inline griptape/tools/email_client/tool.py:18]: Change back to Google docstring.

File: griptape/tools/email_client/tool.py [hunk 4]
```diff
@@ -1,56 +1,70 @@
...
+    """Setting this will override whatever is set as the username parameter for IMAP activities."""
+    imap_password: Optional[str] = field(default=Factory(lambda self: self.password, takes_self=True), kw_only=True)
+    """Setting this will override whatever is set as the username parameter for IMAP activities."""
 
     email_max_retrieve_count: Optional[int] = field(default=None, kw_only=True)
+    """Used to limit the number of messages retrieved during any given activities."""
 
+    # TODO: Ask in PR if mailboxes is unused. Does the mention in the @activity decorator
+    #       provide some functionality or was it was aspirational?
     mailboxes: Optional[dict[str, str]] = field(default=None, kw_only=True)
```
Comment (collindutter) [inline griptape/tools/email_client/tool.py:53]: The description will be rendered in this [jinja template](https://github.com/griptape-ai/griptape/blob/main/griptape/templates/tasks/partials/_tool.j2#L5) that is passed to the LLM. So yeah, `mailboxes` is still used.

**Review by collindutter — CHANGES_REQUESTED** [review]
Overall looks great! Minor feedback items.

File: griptape/tools/email_client/tool.py (hunk 4 above)
Comment (dylanholmes) [reply griptape/tools/email_client/tool.py:49]: - Documented expected usage in the class docstring including an example.
- Added a unit test to verify the mailbox descriptions are rendered in the LLM input. (The unit test was initially just to check my own understanding, but it seems useful, so I included it in the commit).

**Review by collindutter — APPROVED** [review]

File: griptape/loaders/email_loader.py [hunk 5]
```diff
@@ -0,0 +1,80 @@
...
+from griptape.artifacts import ErrorArtifact, ListArtifact, TextArtifact
+from griptape.loaders import BaseLoader
+from typing import Optional
+import imaplib
+import logging
+import mailparser
+
+
+@define(frozen=True)
+class EmailQuery:
```
Comment (vasinov) [inline griptape/loaders/email_loader.py:13]: Can we move this inside `EmailLoader`? Griptape convention is one class per file.

File: griptape/loaders/email_loader.py [hunk 6]
```diff
@@ -0,0 +1,80 @@
...
+    """
+
+    label: str = field(kw_only=True)
+    key: Optional[str] = field(default=None, kw_only=True)
+    search_criteria: Optional[str] = field(default=None, kw_only=True)
+    max_count: Optional[int] = field(default=None, kw_only=True)
+
+@define
+class EmailLoader(BaseLoader):
+    url: str = field(kw_only=True)
```
Comment (vasinov) [inline griptape/loaders/email_loader.py:30]: Could be potentially unclear what URL/protocol is used. Can we rename this to `imap_url`, so it's more descriptive?

File: griptape/tools/email_client/tool.py [hunk 7]
```diff
@@ -1,55 +1,63 @@
...
     smtp_host: Optional[str] = field(default=None, kw_only=True)
     smtp_port: Optional[int] = field(default=None, kw_only=True)
     smtp_use_ssl: bool = field(default=True, kw_only=True)
-
-    # if you set the smtp user/password fields they will override
-    smtp_user: Optional[str] = field(default=None, kw_only=True)
-    smtp_password: Optional[str] = field(default=None, kw_only=True)
-
+    smtp_user: Optional[str] = field(default=Factory(lambda self: self.username, takes_self=True), kw_only=True)
+    smtp_password: Optional[str] = field(default=Factory(lambda self: self.password, takes_self=True), kw_only=True)
```
Comment (vasinov) [inline griptape/tools/email_client/tool.py:45]: Good change!

**Review by vasinov — CHANGES_REQUESTED** [review]
Looks great overall! Could you fix a couple of nits and resolve dependency conflicts with `main` before I merge?

**Review by vasinov — APPROVED** [review]

**Pull request merged**